File: app/src/main/java/com/aiepoissac/busapp/core/data/busservices/DataConnection.java
package com.aiepoissac.busapp.core.data.busservices;

import org.json.simple.JSONObject;
import org.json.simple.parser.JSONParser;
import org.json.simple.parser.ParseException;

import java.io.IOException;
import java.net.HttpURLConnection;
import java.net.URL;
import java.util.ArrayList;
import java.util.Scanner;

final class DataConnection {

    static ArrayList<JSONObject> getData(String description, int count) {
        try {
            URL url = null;
            if (description.equals("BUS STOPS")) {
                url = new URL("https://datamall2.mytransport.sg/ltaodataservice/BusStops?$skip=" + count);
            } else if (description.equals("BUS ROUTES")) {
                url = new URL("https://datamall2.mytransport.sg/ltaodataservice/BusRoutes?$skip=" + count);
            } else if (description.equals("BUS SERVICES")) {
                url = new URL("https://datamall2.mytransport.sg/ltaodataservice/BusServices?$skip=" + count);
            }
            JSONObject jsonObject = getJsonObject(url);
            @SuppressWarnings("unchecked")
            ArrayList<JSONObject> jsonArray = (ArrayList<JSONObject>) jsonObject.get("value");
            return jsonArray;
        } catch (IOException | ParseException | NullPointerException e) {
            return new ArrayList<>();
        }
    }

    private static JSONObject getJsonObject(URL url) throws IOException, ParseException {
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();
        connection.setRequestMethod("GET");
        connection.setRequestProperty("AccountKey", "***REMOVED***");
        connection.setRequestProperty("accept", "application/json");
        connection.connect();
        StringBuilder result = new StringBuilder();
        Scanner scanner = new Scanner(connection.getInputStream());
        while (scanner.hasNext()) {
            result.append(scanner.nextLine());
        }
        scanner.close();
        JSONParser parser = new JSONParser();
        return (JSONObject) parser.parse(result.toString());
    }

}

File: app/src/main/java/com/aiepoissac/busapp/data/businfo/DataConnection.kt
package com.aiepoissac.busapp.data.businfo

import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext
import kotlinx.serialization.json.Json
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL
import java.util.Scanner


enum class BusDataType {
    BusServices,
    BusRoutes,
    BusStops
}

private suspend fun getData(dataType: BusDataType, count: Int): String = withContext(Dispatchers.IO) {
    val url: URL?
    if (dataType == BusDataType.BusStops) {
        url = URL("https://datamall2.mytransport.sg/ltaodataservice/BusStops?\$skip=$count")
    } else if (dataType == BusDataType.BusRoutes) {
        url = URL("https://datamall2.mytransport.sg/ltaodataservice/BusRoutes?\$skip=$count")
    } else if (dataType == BusDataType.BusServices) {
        url = URL("https://datamall2.mytransport.sg/ltaodataservice/BusServices?\$skip=$count")
    } else throw IllegalArgumentException()
    val connection = url.openConnection() as HttpURLConnection
    connection.requestMethod = "GET"
    connection.setRequestProperty("AccountKey", "***REMOVED***")
    connection.setRequestProperty("accept", "application/json")
    connection.connect()
    if (connection.responseCode != 200) {
        throw IOException(connection.responseMessage)
    } else {
        val result = StringBuilder()
        val scanner = Scanner(connection.inputStream)
        while (scanner.hasNext()) {
            result.append(scanner.nextLine())
        }
        scanner.close()
        return@withContext result.toString()
    }
}


suspend fun populateBusServices(busRepository: BusRepository) {
    busRepository.deleteAllBusServices()
    var i = 0
    while (true) {
        val json = withContext(Dispatchers.IO) {
            getData(dataType = BusDataType.BusServices, count = i * 500)
        }
        val busServicesInfo: BusServicesInfo = Json.decodeFromString<BusServicesInfo>(json)
        if (busServicesInfo.value.isEmpty()) {
            break
        } else {
            for (busServiceInfo in busServicesInfo.value) {
                busRepository.insertBusService(busServiceInfo)
            }
        }
        i++
    }
}

suspend fun populateBusRoutes(busRepository: BusRepository) {
    busRepository.deleteAllBusRoutes()
    var i = 0
    while (true) {
        val json = withContext(Dispatchers.IO) {
            getData(dataType = BusDataType.BusRoutes, count = i * 500)
        }
        val busRoutesInfo: BusRoutesInfo = Json.decodeFromString<BusRoutesInfo>(json)
        if (busRoutesInfo.value.isEmpty()) {
            break
        } else {
            for (busRouteInfo in busRoutesInfo.value) {
                busRepository.insertBusRoute(busRouteInfo)
            }
        }
        i++
    }
}

suspend fun populateBusStops(busRepository: BusRepository) {
    busRepository.deleteAllBusStops()
    var i = 0
    while (true) {
        val json = withContext(Dispatchers.IO) {
            getData(dataType = BusDataType.BusStops, count = i * 500)
        }
        val busStopsInfo: BusStopsInfo = Json.decodeFromString(json)
        if (busStopsInfo.value.isEmpty()) {
            break
        } else {
            for (busStopInfo in busStopsInfo.value) {
                busRepository.insertBusStop(busStopInfo)
            }
        }
        i++
    }
}